### PR TITLE
Begin immutable refactoring

### LINKED ISSUE #8310: Type with templated closure is not inferred correctly
Turns out I actually fixed this bug in my immutable refactoring, couldn't initially figure out why was upstream returning this weird result :)
https://psalm.dev/r/a2a2c35578

### PR DESCRIPTION
**Pull request by danog**
As per #8132, fixes #8310

### CODE REVIEW

**Comment by danog** [comment]
Settling on a simpler structure based on ImmutableUnions to avoid churn.

File: src/Psalm/Internal/Algebra/FormulaGenerator.php [hunk 1]
```diff
@@ -153,7 +153,6 @@ public static function getFormula(
                         $redefined = false;
 
                         if ($var[0] === '=') {
-                            /** @var string */
```
Comment (orklah) [inline src/Psalm/Internal/Algebra/FormulaGenerator.php:156]: This is only possible because you set composer to PHP 8.1. We decided that Psalm must still be able to run on PHP 7.4 for its version 5.

You should revert the change on composer to still allow 7.4, otherwise this won't be merged in Psalm 5 (and will probably won't be released soon: allowing only PHP8.1 would mean denying two third of our users)

Comment (danog) [reply src/Psalm/Internal/Algebra/FormulaGenerator.php:156]: I'll revert this, but still, I see nothing wrong with restricting psalm to 8.1, since it's a development tool that is only ever required as a development dependency and runs mainly in CI jobs, where the PHP version isn't as important and doesn't require changes to production code.

Comment (orklah) [reply src/Psalm/Internal/Algebra/FormulaGenerator.php:156]: I don't have stats for that, but there is a lot of Psalm runs that are made manually by developpers locally. Once you take that into account, you have to factor in every developer config. For example, for a dev on windows running Wamp to run their application on their laptop, installing a new PHP version just for Psalm may be a real pain.

Psalm may also need to use reflection on particular extensions that may only be available for certain PHP versions. 

We made a big leap to 7.4 for Psalm 5 (remember Psalm 4 supported PHP 7.1. It was initially 7.3 but users complained). It still represents about 6% of Psalm installs to this day that we're gonna left behind. We feel it's acceptable, but we're not ready to dump 66% (not even if we could say that they can easily change their CI jobs)

Comment (danog) [reply src/Psalm/Internal/Algebra/FormulaGenerator.php:156]: That's understandable (especially the windows part), I'll revert this here and keep using 8.1 in my own fork, I really want to use readonly properties to catch eventual runtime corruptions :)

**Comment by danog** [comment]
Almost done here, the only strange thing left is the assertion testcase failure, they're caused by https://github.com/vimeo/psalm/blob/master/src/Psalm/Internal/Analyzer/Statements/Expression/CallAnalyzer.php#L766, which relies on an inconsistent internal state of `Union` after template substitution, which the old code caused (the `types` property is correctly populated but the `literal_*_types` properties are empty), it's really strange that we're relying on this behavior and I'm not really sure what to do here, any help? :)

**Comment by danog** [comment]
Actually hang on, it seems like the `allowCanBeNotSameAfterAssertionScalar` unit test itself is wrong: 
```php
<?php
                    namespace Bar;

                    /**
                     * Asserts that two variables are the same.
                     *
                     * @template T
                     * @param T      $expected
                     * @param mixed  $actual
                     * @psalm-assert !=T $actual
                     */
                    function assertNotSame($expected, $actual) : void {}

                    $c = 4;
                    $d = rand(0, 1) ? 4 : 5;
                    assertNotSame($d, $c);

                    function foo(string $a, string $b) : void {
                        assertNotSame($a, $b);
                    }
```

It's basically asserting that `$c != $d`, aka `4 != 4|5` aka `4 != 4 || 4 != 5`, and the first assertion is indeed impossible

**Comment by danog** [comment]
The other unit test failure seems actually like a (partial) improvement over the previous behavior, so I'm actually thinking of correcting the testcase to account for the new behavior

**Comment by orklah** [comment]
I don't get how the unit test is wrong. You should be able to replace `assertNotSame($d, $c);` by `assert($d !== $c);`. It's a valid thing to check

**Comment by danog** [comment]
> I don't get how the unit test is wrong. You should be able to replace `assertNotSame($d, $c);` by `assert($d !== $c);`. It's a valid thing to check

I feel there is a slight distinction (https://psalm.dev/r/05628e8554):
- `assert($expected !== $actual)`: psalm errors if there is no intersection between the types of  `$actual` and `$expected`
- `assertNotSame($expected, $actual)` psalm errors if the type of `$actual` doesn't contain the type of `$expected`

Or at least, that's the current behavior with this MR.
The previous behavior was:
- `assert($expected !== $actual)`: psalm errors if there is no intersection between the types of  `$actual` and `$expected`
- `assertNotSame($expected, $actual)`:
  - if `$expected` is a union type, psalm errors if there is no intersection between the types of  `$actual` and `$expected`
  - otherwise, psalm errors if the type of `$actual` doesn't contain the type of `$expected`

I'm not sure what should be the correct behavior, given that to obtain the previous behavior we relied on an inconsistent state of union, possibly caused by a mistake/bug.

Could anyone clarify on what was the actual, intended, behavior, and if it relied on that inconsistent state of Union?

**Comment by danog** [comment]
I think this is what was @boesing actually meant when using `isSingle`, fixed the testcase.

**Comment by boesing** [comment]
Just as a sidenote, AFAIR the isSingle check was there before I added my stuff. I just refactored it a bit. Probably worth going through the history to find the initial commit which added the check.

I was super confused by these test cases as well when I worked on adding my code. I somehow managed to change the code so it works for the old test cases and my new code.

**Comment by danog** [comment]
This should do the trick nicely I think :)

**Comment by danog** [comment]
All done!

**Comment by danog** [comment]
I'll continue refactoring in a separate MR, I still have to implement atomic immutability (at least for nested Union properties) and switch to @readonly properties in storages.

File: src/Psalm/Internal/Analyzer/Statements/Expression/Assignment/ArrayAssignmentAnalyzer.php [hunk 2]
```diff
@@ -742,17 +742,21 @@ private static function analyzeNestedArrayAssignment(
 
             $is_last = $i === count($child_stmts) - 1;
 
+            $temp = $child_stmt_dim_type ?? Type::getInt();
```
Comment (orklah) [inline src/Psalm/Internal/Analyzer/Statements/Expression/Assignment/ArrayAssignmentAnalyzer.php:745]: Could you change the variable in something more descriptive please?

File: src/Psalm/Internal/Analyzer/Statements/Expression/BitwiseNotAnalyzer.php [hunk 3]
```diff
@@ -59,6 +60,7 @@ public static function analyze(
                         new TLiteralInt(~$type_part->value) :
                         new TInt;
 
+                    // todo:immutable this makes no sense?
```
Comment (orklah) [inline src/Psalm/Internal/Analyzer/Statements/Expression/BitwiseNotAnalyzer.php:63]: is this a part of what you intend to fix in a next PR?

**Review by orklah — COMMENTED** [review]
A few things:
- do we need bustCache in Union still? IIRC, it's used to rerender the internal id when the union changed. It shouldn't do that anymore.
- Could we make a check for all `clone $union` everywhere and get rid of them? it shouldn't be needed anymore
- Could you document this change in README.md. It's actually a big change for plugin authors and they'll need to be aware of this

@weirdan @muglug @AndrolGenhald this is a big fundamental change. I could use your help reviewing this :)

**Comment by AndrolGenhald** [comment]
> do we need bustCache in Union still? IIRC, it's used to rerender the internal id when the union changed. It shouldn't do that anymore.

The scope of this PR is actually smaller than I expected. The `Atomic`s in the `Union` are mutable and can still change so that'll still be required when that happens. Some calls might be removable now though, if they haven't been already.

> Could we make a check for all clone $union everywhere and get rid of them? it shouldn't be needed anymore

It's probably still needed in most cases since `Atomic`s are still mutable.

**Comment by orklah** [comment]
Oh right, didn't thought about that

File: src/Psalm/Internal/Analyzer/Statements/Expression/BitwiseNotAnalyzer.php (hunk 3 above)
Comment (danog) [reply src/Psalm/Internal/Analyzer/Statements/Expression/BitwiseNotAnalyzer.php]: Actually now that I look at it, that code makes perfect sense, removed the comment :)

**Comment by danog** [comment]
>do we need bustCache in Union still? IIRC, it's used to rerender the internal id when the union changed. It shouldn't do that anymore.

As atomics are still mutable (they won't be in my next PR), bustCache still has some possible uses.

>Could we make a check for all clone $union everywhere and get rid of them? it shouldn't be needed anymore

Same thing about atomics + even after immutable atomics, there __might__ be some usecases to cloning types, since I'm still not sure whether to mark as `@readonly` **all** properties of Unions and Atomics (not just those with type Union/Atomic).

>Could you document this change in README.md. It's actually a big change for plugin authors and they'll need to be aware of this

Done!

**Comment by orklah** [comment]
Seems great. Let's wait a few days to see if Matt or Bruce have time to chime in and we'll be good to go :)

Thanks for the massive work!

**Comment by orklah** [comment]
Alright, looks like we didn't get more comments.

I'm ready to merge, I just need to make sure you're ready to do the last part (otherwise, it won't make much sense as is) :)

**Comment by orklah** [comment]
I wonder if this is a change big enough to consider delivering V5 soon and merge for a V6 that we would start immediately after.

All things considered, we probably should have released V5 long ago and started a V6 already. Given how we lack long term vision, we should probably deliver major version more often to avoid piling big changes together...

**Comment by AndrolGenhald** [comment]
> I wonder if this is a change big enough to consider delivering V5 soon and merge for a V6 that we would start immediately after.

FWIW I was hoping to get a V6 relatively soon after V5 anyway (or at least sooner than V5 is after V4) with the major type system changes I've been wanting to do (and never managed to get fully working), but it's probably not going to be until next year till that I can devote a lot of time to that again.

**Comment by danog** [comment]
>I'm ready to merge, I just need to make sure you're ready to do the last part (otherwise, it won't make much sense as is) :)

I'm actually already working on it on the immutable_readonly branch!

**Comment by orklah** [comment]
Yeah, September is as good a date as any, we just need to lock a date and keep it. I'll leave this unmerged for now, it may be hard to merge incoming 4.x changes in the meantime otherwise

**Comment by muglug** [comment]
I think this is a fine v5 change if you want to target September. Sorry I haven’t been more active, but I wanted to say earlier: when it comes to these sorts of big changes, I rely on Psalm and Psalm’s test suite to detect regressions.

Over Psalm’s history I have done many big refactors, and very few of them broke some downstream consumer’s pipeline.

**Comment by orklah** [comment]
Thanks for feedback Matt! Did you had issues with plugin authors on previous majors?

This is definitely a change that will affect them directly

**Comment by muglug** [comment]
No — I think if the intent of the refactoring is sound, people generally “get it”
